core/src/services/memcached/binary.rs

```diff
@@ -90,26 +95,135 @@ impl PacketHeader {
+            Self::Tls(conn) => conn.get(key).await,
+            Self::Tcp(conn) => conn.get(key).await,
+        }
+    }
+
+    pub async fn set(&mut self, key: &str, val: &[u8], expiration: u32) -> Result<()> {
+        match self {
+            Self::Tls(conn) => conn.set(key, val, expiration).await,
+            Self::Tcp(conn) => conn.set(key, val, expiration).await,
+        }
+    }
+    pub async fn delete(&mut self, key: &str) -> Result<()> {
+        match self {
+            Self::Tls(conn) => conn.delete(key).await,
+            Self::Tcp(conn) => conn.delete(key).await,
+        }
+    }
+}
+
 pub struct Response {
     header: PacketHeader,
     _key: Vec<u8>,
     _extras: Vec<u8>,
     value: Vec<u8>,
 }
 
-pub struct Connection {
+pub struct TlsConnection {
+    io: BufReader<TlsStream<TcpStream>>,
+}
+
+impl TlsConnection {
+    pub fn new(io: TlsStream<TcpStream>) -> Self {
+        Self {
+            io: BufReader::new(io),
+        }
+    }
+}
+
+impl Conn for TlsConnection {
+    type T = TlsStream<TcpStream>;
+    fn get_conn(&mut self) -> &mut Self::T {
+        self.io.get_mut()
+    }
+}
+pub struct TcpConnection {
     io: BufReader<TcpStream>,
 }
 
-impl Connection {
+impl TcpConnection {
     pub fn new(io: TcpStream) -> Self {
         Self {
             io: BufReader::new(io),
         }
     }
+}
+
+impl Conn for TcpConnection {
+    type T = TcpStream;
+    fn get_conn(&mut self) -> &mut Self::T {
+        self.io.get_mut()
+    }
+}
+
+pub async fn parse_response<T: AsyncWriteExt + std::marker::Unpin + tokio::io::AsyncRead>(
+    reader: &mut T,
+) -> Result<Response> {
+    let header = PacketHeader::read::<T>(reader)
+        .await
+        .map_err(new_std_io_error)?;
+
+    if header.vbucket_id_or_status != constants::OK_STATUS
+        && header.vbucket_id_or_status != constants::KEY_NOT_FOUND
+    {
+        return Err(
+            Error::new(ErrorKind::Unexpected, "unexpected status received")
+                .with_context("message", format!("{}", header.vbucket_id_or_status)),
+        );
+    }
+
+    let mut extras = vec![0x0; header.extras_length as usize];
+    reader
+        .read_exact(extras.as_mut_slice())
+        .await
+        .map_err(new_std_io_error)?;
+
+    let mut key = vec![0x0; header.key_length as usize];
+    reader
+        .read_exact(key.as_mut_slice())
+        .await
+        .map_err(new_std_io_error)?;
 
-    pub async fn auth(&mut self, username: &str, password: &str) -> Result<()> {
-        let writer = self.io.get_mut();
+    let mut value = vec![
+        0x0;
+        (header.total_body_length - u32::from(header.key_length) - u32::from(header.extras_length))
+            as usize
+    ];
+    reader
+        .read_exact(value.as_mut_slice())
+        .await
+        .map_err(new_std_io_error)?;
+
+    Ok(Response {
+        header,
+        _key: key,
+        _extras: extras,
+        value,
+    })
+}
+#[async_trait::async_trait]
+pub trait Conn {
```

Comment:
This trait is unnecessary since we only have two variants. We can use an enum to represent it.